### PR TITLE
sip MESSAGE method in camel-sip

### PR DESCRIPTION
**Pull request by nikvaessen**
sip endpoints can now retrieve or send sip messages

### CODE REVIEW

**Comment by davsclaus** [comment]
Can you create a JIRA ticket about this. And is there more work needed in this PR. You have done many code changes so its harder to review.

**Comment by nikvaessen** [comment]
The issue in jira is here: https://issues.apache.org/jira/browse/CAMEL-9190

Can you give me some direction as to were to put more work in? 

I tested my code by creating two camel routes which can be seen here:
https://github.com/nikvaessen/camel-sip-test/blob/sip_messaging/src/main/java/org/opentelecoms/camelsiptest/RouteBuilder.java

One routes listens for incoming messages. When someone send a sip message to the address "listener@localhost:5154", it gets stored in an activeMQ queue.
The other route gets every message from the queue and sends it to the address I specified, in this case the same sip address I'm sending it from.
When the producer processes the exchange, it looks for a header to know what what kind of SIP method the requests needs to be send with. So I made a procesesor in the route which gives it the right header, so the existing code can already send it. I do not have an idea how to change it so the producer knows which header to use..

File: components/camel-sip/pom.xml [hunk 1]
```diff
@@ -26,6 +26,7 @@
   </parent>
   
 	<artifactId>camel-sip</artifactId>
+	<version>2.18-SNAPSHOT</version>
```
Comment (dpocock) [inline components/camel-sip/pom.xml]: does this need to be here?

File: components/camel-sip/src/main/docs/sip.adoc [hunk 2]
```diff
@@ -80,8 +80,13 @@ The SIP component has no options.
 
 
 
+
+
+
+
+
```
Comment (dpocock) [inline components/camel-sip/src/main/docs/sip.adoc]: blank lines could be removed

File: components/camel-sip/src/main/docs/sip.adoc [hunk 3]
```diff
@@ -140,6 +146,7 @@ The SIP component supports 45 endpoint options which are listed below:
 Sending Messages to/from a SIP endpoint
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+
```
Comment (dpocock) [inline components/camel-sip/src/main/docs/sip.adoc]: is this blank line consistent with the rest of the spacing in the document?

**Comment by dpocock** [comment]
@nikvaessen can you please log in to Jira and add a comment to CAMEL-9190 with a link to this pull request?  I agree with the comment from @davsclaus that it is a bit difficult to review in Github, I wonder if it looks better when viewed with ```gitk``` though, it has an option to ignore whitespace changes.

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java [hunk 4]
```diff
@@ -53,124 +53,348 @@
...
     private Map<String, Object> parameters;
 
+    /**
+     * The SIP URI the SipEndpoint needs to connect to.
+     * This object represents the SIP URI string given when the SipEndpoint is created.
+     */
     @UriPath @Metadata(required = "true")
     private URI uri;
-    @UriParam(label = "advanced")
-    private AddressFactory addressFactory;
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java]: @nikvaessen is there a reason you changed the order of these variables?  Could you put these variables back in the same order so the diff is easier to review?  Thanks for inserting the comments about each of them.

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java [hunk 5]
```diff
@@ -53,124 +53,348 @@
...
+    /**
+     * The name of the SipStack. Defaults to "NAME_NOT_SET".
+     */
     @UriParam(label = "common", defaultValue = "NAME_NOT_SET")
     private String stackName = "NAME_NOT_SET";
-    @UriParam(label = "common", defaultValue = "tcp", enums = "tcp,udp")
+
+    /**
+     * The transport method used to send and receive messages. Defaults to TCP.
+     * Has to be equal to TCP or UDP to be valid.
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java]: @nikvaessen please mention TLS in the comment too

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java [hunk 6]
```diff
@@ -53,124 +53,348 @@
...
+     */
+    @UriParam(label = "consumer")
+    private boolean consumer;
+
+    /**
+     * Whether the sip consumer will subscribe to the given address or only listen for incoming
+     * MESSAGE requests
+     */
+    @UriParam(label = "consumer", defaultValue = "false")
+    private boolean subscribing = false;
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java]: @nikvaessen is this default value consistent with the existing behavior of the component?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java [hunk 7]
```diff
@@ -243,32 +483,40 @@ public void parseURI() throws Exception {
         LOG.trace("Consumer:" + consumer + " StackName:" + stackName);
         LOG.trace("From User: " + getFromUser() + " From host: " + getFromHost() + " From Port: " + getFromPort());
          
-        createFactoriesAndHeaders(parameters, component);
-        
+        this.createFactoriesAndHeaders(parameters, component);
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java]: @nikvaessen why did you add "this" here?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java [hunk 8]
```diff
@@ -243,32 +483,40 @@ public void parseURI() throws Exception {
...
+     * @param parameters the parameters given when endpoint got created
+     * @param component the SipComponent
+     * @throws Exception when something goes wrong looking up from the parameters map
+     */
     @SuppressWarnings("unchecked")
     private void createFactoriesAndHeaders(Map<String, Object> parameters, SipComponent component) throws Exception {
         headerFactory = sipFactory.createHeaderFactory();
         addressFactory = sipFactory.createAddressFactory();
-        setMessageFactory(sipFactory.createMessageFactory());
+        this.setMessageFactory(sipFactory.createMessageFactory());
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java]: @nikvaessen why did you add "this" here?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java [hunk 9]
```diff
@@ -243,32 +483,40 @@ public void parseURI() throws Exception {
...
         
         fromHeader = component.resolveAndRemoveReferenceParameter(parameters, "fromHeader", FromHeader.class, null);
         if (fromHeader == null) { 
             createFromHeader();
         }
-        if (!presenceAgent) {
+        if(!consumer || subscribing) {
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java]: @nikvaessen you use this expression in several places, could you write a helper function that returns a boolean value to encapsulate this expression?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java [hunk 10]
```diff
@@ -389,11 +699,24 @@ private void createContactHeader() throws ParseException {
...
-    Properties createInitialProperties() {
+
+    /**
+     * creates a properties object holding information about the stack name, the limit on message size,
+     * whether connections should be cashed, if messages can be send via proxies and optionally where the logs
+     * are located
+     *
+     * @return a properties object holding the information stated above
+     */
+    public Properties createInitialProperties() {
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java]: @nikvaessen does this absolutely need to become public?  @davsclaus is there any other pattern you would prefer to see used instead of making this method public?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java [hunk 11]
```diff
@@ -408,19 +731,15 @@ Properties createInitialProperties() {
         return properties;
     }
 
-    public AddressFactory getAddressFactory() {
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConfiguration.java]: @nikvaessen many of these getter methods have been re-ordered, this makes the diff much longer than necessary, could you make another commit putting these methods back in the original order?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConsumer.java [hunk 12]
```diff
@@ -0,0 +1,203 @@
+/**
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipConsumer.java]: this appears to be a rename from SipSubscriber.java, its a shame Github doesn't show it more concisely though

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/SipPublisher.java [hunk 13]
```diff
@@ -80,10 +81,10 @@ protected void doStop() throws Exception {
         getSipStack().stop();
     }
     
-    public void process(Exchange exchange) throws Exception {
+    public void process(Exchange exchange) throws Exception { //// TODO: 30/06/16 fix
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/SipPublisher.java]: @nikvaessen what is the TODO on this line?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java [hunk 14]
```diff
@@ -85,21 +93,21 @@ private void sendNotification(EventHeader eventHeader, boolean isInitial, Object
         ((SipURI)sipPresenceAgent.getConfiguration().getContactHeader().getAddress().getURI()).setParameter(
                 sipPresenceAgent.getConfiguration().getFromUser(), sipPresenceAgent.getConfiguration().getFromHost());
 
-        SubscriptionStateHeader sstate;
+        SubscriptionStateHeader state;
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java]: @nikvaessen the original name, ```sstate``` appears to be an abbreviation of Subscription State, not a typo.  Could you rename this variable back to ```sstate```?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java [hunk 15]
```diff
@@ -179,6 +187,64 @@ public void processSubscribe(RequestEvent requestEvent,
         }
     }
 
+    public void processMessage(RequestEvent requestEvent, ServerTransaction serverTransactionId) {
+        Request request = requestEvent.getRequest();
+        try
+        {
+            LOG.debug("SipPresenceAgentListener: Received an MESSAGE request");
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java]: @nikvaessen are you sure it is always ```MESSAGE``` at this point?  Maybe check the request method == ```MESSAGE``` and log an error if something else comes in.

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java [hunk 16]
```diff
@@ -179,6 +187,64 @@ public void processSubscribe(RequestEvent requestEvent,
...
+            SipProvider sipProvider = (SipProvider) requestEvent.getSource();
+            String contentType = request.getHeader("Content-Type").toString();
+
+            if(contentType == null)
+            {
+                response = messageFactory.createResponse(400, request); // bad request
+                sipProvider.sendResponse(response);
+                return;
+            }
+            else if(contentType.contains("text/plain"))
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java]: @nikvaessen could you allow other content types too?  They can be accepted by Camel as binary data (byte[]), just make sure you store the content type name in the header ```Exchange.CONTENT_TYPE```

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java [hunk 17]
```diff
@@ -46,22 +50,26 @@
     protected Dialog dialog;
     protected int notifyCount;
     private SipPresenceAgent sipPresenceAgent;
+    private Queue<String> messageQueue;
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java]: @nikvaessen could you store whole SIP requests in this Queue, rather than just the String body?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java [hunk 18]
```diff
@@ -179,6 +187,64 @@ public void processSubscribe(RequestEvent requestEvent,
...
+            LOG.debug("Exception thrown during processing Message request in SipPresenceAgentListener");
+        }
+    }
+
+    public boolean hasRetrievedMessage()
+    {
+        return !messageQueue.isEmpty();
+    }
+
+    public String getRetrievedMessage() throws NoSuchElementException
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipPresenceAgentListener.java]: @nikvaessen instead of just returning the message body as a string, could you create an Exchange object and populate the "in" message body and various headers?

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipSubscriptionListener.java [hunk 19]
```diff
@@ -34,48 +35,106 @@
...
         LOG.debug("Consumer Dispatching the received notification along the route");
-        Exchange exchange = sipSubscriber.getEndpoint().createExchange(ExchangePattern.InOnly);
+        Exchange exchange = sipConsumer.getEndpoint().createExchange(ExchangePattern.InOnly);
         exchange.getIn().setBody(response);
-        try {
-            sipSubscriber.getProcessor().process(exchange);
-        } catch (Exception e) {
+
+        //try to send the message along
+        try
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipSubscriptionListener.java]: @nikvaessen it looks like you moved the ```{``` to the next line, please use checkstyle to catch things like this, it is described here:
http://camel.apache.org/building.html

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipSubscriptionListener.java [hunk 20]
```diff
@@ -34,48 +35,106 @@
...
+     * process incoming SIP requests. Only processes NOTIFY and MESSAGE requests
+     * @param requestReceivedEvent the request event
+     */
     public void processRequest(RequestEvent requestReceivedEvent) {
         Request request = requestReceivedEvent.getRequest();
         ServerTransaction serverTransactionId = requestReceivedEvent
                 .getServerTransaction();
         String viaBranch = ((ViaHeader)(request.getHeaders(ViaHeader.NAME).next())).getParameter("branch");
-        LOG.debug("Request: {}", request.getMethod()); 
+        String fromBranch = ((FromHeader)(request.getHeader(FromHeader.NAME))).toString();
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipSubscriptionListener.java]: @nikvaessen this should just be ```from```, not ```fromBranch```

File: components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipSubscriptionListener.java [hunk 21]
```diff
@@ -34,48 +35,106 @@
...
-        if (request.getMethod().equals(Request.NOTIFY)) {
+        //process NOTIFY and MESSAGE requests only
+        if (Request.NOTIFY.equals(request.getMethod()))
+        {
             processNotify(requestReceivedEvent, serverTransactionId);
-        } 
+        }
+        else if(Request.MESSAGE.equals(request.getMethod()))
+        {
+            processMessage(requestReceivedEvent, serverTransactionId);
```
Comment (dpocock) [inline components/camel-sip/src/main/java/org/apache/camel/component/sip/listener/SipSubscriptionListener.java]: @nikvaessen as it is SIP "SIMPLE", maybe this method should be called ```processSimpleMessage``` to make it more obvious what it is

**Comment by dpocock** [comment]
@nikvaessen would you be able to review the latest comments I made through Github and make more commits fixing them?

After that, could you please ask @davsclaus for his opinion on the diff?

Overall, it looks like some of the whitespace changes you made were there to satisfy checkstyle and this makes the commit look bigger.  Adding comments is completely OK.  Moving methods around in one of the classes though, without changing them in any way, is probably not a good idea as it just makes it harder to review.

I think you also need to look closely at how you encapsulate the messages coming in from SIP, you need to store the "From" and "To" values from the SIP request as headers on the Camel Exchange that is created.  This isn't a difficult thing to fix.

**Comment by davsclaus** [comment]
This PR is hard to review. Are you able to resubmit the PR in a new PR and we can start over again.

Also please log a JIRA ticket about this enhancement.

**Comment by nikvaessen** [comment]
I will be able to resubmit this PR over the christmas break
